File: src/fancywords/FancyWordsDatabase.java
package fancywords;

import com.zaxxer.hikari.HikariDataSource;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

/**
 *
 */
public class FancyWordsDatabase implements AutoCloseable {

    final static int RECOMMENDED_SIZE = 2 * Runtime.getRuntime().availableProcessors() + 1;
    private final static String DATABASE_NAME = "fancy_words_server", ALLOW_MULTI_QUERIES_TRUE = "allowMultiQueries=true";
    private final static String URL = "jdbc:mariadb://localhost:3306/" + DATABASE_NAME + "?" + ALLOW_MULTI_QUERIES_TRUE,
            USERNAME = "fancywordsserver",
            PASSWORD = "xhqycqyys";
    HikariDataSource hikariDataSource;

    public FancyWordsDatabase() throws SQLException {
        hikariDataSource = new HikariDataSource();
        hikariDataSource.setDriverClassName("org.mariadb.jdbc.Driver");
        //hikariDataSource.setDataSourceClassName("org.mariadb.jdbc.MariaDbDataSource");
        hikariDataSource.setJdbcUrl(URL);
        hikariDataSource.setUsername(USERNAME);
        hikariDataSource.setPassword(PASSWORD);

        hikariDataSource.setMaximumPoolSize(RECOMMENDED_SIZE);

        hikariDataSource.addDataSourceProperty("prepStmtCacheSize", 256);
        hikariDataSource.addDataSourceProperty("prepStmtCacheSqlLimit", 2048);
        hikariDataSource.addDataSourceProperty("cachePrepStmts", true);
        hikariDataSource.addDataSourceProperty("useServerPrepStmts", true);
    }

    public void close() {
        hikariDataSource.close();
    }

    public boolean authenticateUser(String username, byte[] encryptedPassword) {
        // TODO implement here
        return false;
    }

    public List<BriefUserWordEntry> queryAllEntries() throws SQLException {
        try (Connection connection = hikariDataSource.getConnection()) {
            PreparedStatement preparedStatement = connection.prepareStatement("SELECT word, definition, log_absolute_mastery FROM user_word_entry ORDER BY word ASC;");

            ResultSet resultSet = preparedStatement.executeQuery();
            ArrayList<BriefUserWordEntry> entries = new ArrayList<>();
            for (int i = 0; resultSet.next(); i++) {
                entries.add(new BriefUserWordEntry(resultSet.getString(1),
                        resultSet.getString(2),
                        resultSet.getDouble(3)));
            }

            connection.close();

            return entries;
        } catch (SQLException e) {
            throw e;
        }
    }

    public boolean insertUserWordEntry(UserWordEntry userWordEntry) throws SQLException {
        try (Connection connection = hikariDataSource.getConnection()) {
            PreparedStatement preparedStatement = connection.prepareStatement("INSERT INTO user_word_entry VALUES (?, ?, ?, ?, ?, ?, ?, ?);");
            preparedStatement.setString(1, userWordEntry.getWord());
            preparedStatement.setString(2, userWordEntry.getDefinition());
            preparedStatement.setDouble(3, userWordEntry.getLogAbsoluteMastery());
            preparedStatement.setLong(4, userWordEntry.getTimeCreated());
            preparedStatement.setLong(5, userWordEntry.getTimeEdited());
            preparedStatement.setInt(6, userWordEntry.getCountViewed());
            preparedStatement.setInt(7, userWordEntry.getCountTested());
            preparedStatement.setInt(8, userWordEntry.getCountTestedCorrect());
            long updateCount;
            try {
                updateCount = preparedStatement.executeUpdate();
            } catch (SQLException e) {
                if (e.getErrorCode() == 1062)
                    return false;
                else throw e;
            }

            connection.close();

            return updateCount > 0;
        } catch (SQLException e) {
            throw e;
        }
    }

    public boolean updateDefinition(String wordToBeUpdated, String definition) throws SQLException {
        try (Connection connection = hikariDataSource.getConnection()) {
            PreparedStatement preparedStatement = connection.prepareStatement("UPDATE user_word_entry SET definition = ? WHERE word = ?;");
            preparedStatement.setString(1, definition);
            preparedStatement.setString(2, wordToBeUpdated);

            long updateCount = preparedStatement.executeUpdate();

            connection.close();

            return updateCount > 0;
        } catch (SQLException e) {
            throw e;
        }
    }

    public boolean updateMasteryByOne(String wordToBeUpdated, long time) throws SQLException {
        try (Connection connection = hikariDataSource.getConnection()) {
            connection.setAutoCommit(false);
            PreparedStatement selectPreparedStatement = connection.prepareStatement("SELECT log_absolute_mastery FROM user_word_entry WHERE word = ?;");
            selectPreparedStatement.setString(1, wordToBeUpdated);
            ResultSet resultSet = selectPreparedStatement.executeQuery();
            double logAbsoluteMastery;
            if (resultSet.next())
                logAbsoluteMastery = resultSet.getDouble(1);
            else
                return false;

            PreparedStatement updatePreparedStatement = connection.prepareStatement("UPDATE user_word_entry SET log_absolute_mastery = ? WHERE word = ?;");
            updatePreparedStatement.setDouble(1, Utils.increaseLogAbsoluteMasteryByOneByTime(logAbsoluteMastery, time));
            updatePreparedStatement.setString(2, wordToBeUpdated);

            long updateCount = updatePreparedStatement.executeUpdate();

            connection.commit();
            connection.setAutoCommit(true);

            connection.close();

            return updateCount > 0;
        } catch (SQLException e) {
            throw e;
        }
    }

    public boolean updateTimeEdited(String wordToBeUpdated, long timeEdited) throws SQLException {
        try (Connection connection = hikariDataSource.getConnection()) {
            PreparedStatement preparedStatement = connection.prepareStatement("UPDATE user_word_entry SET time_edited = ? WHERE word = ?;");
            preparedStatement.setLong(1, timeEdited);
            preparedStatement.setString(2, wordToBeUpdated);

            long updateCount = preparedStatement.executeUpdate();

            connection.close();

            return updateCount > 0;
        } catch (SQLException e) {
            throw e;
        }
    }

    /**
     * This method automatically increase the number of times this word has been viewed by ONE.
     *
     * @param wordToBeUpdated
     * @return
     * @throws SQLException
     */
    public boolean updateCountViewedByOne(String wordToBeUpdated) throws SQLException {
        try (Connection connection = hikariDataSource.getConnection()) {
            PreparedStatement preparedStatement = connection.prepareStatement("UPDATE user_word_entry SET count_viewed = count_viewed + 1 WHERE word = ?;");
            preparedStatement.setString(1, wordToBeUpdated);

            long updateCount = preparedStatement.executeUpdate();

            connection.close();

            return updateCount > 0;
        } catch (SQLException e) {
            throw e;
        }
    }

    public boolean updateCountTested(String wordToBeUpdated, boolean correct) throws SQLException {
        try (Connection connection = hikariDataSource.getConnection()) {
            PreparedStatement preparedStatement = connection.prepareStatement(correct ?
                    "UPDATE user_word_entry SET count_tested = count_tested + 1 AND count_tested_correct = count_tested_correct + 1 WHERE word = ?;" :
                    "UPDATE user_word_entry SET count_tested = count_tested + 1 WHERE word = ?;");
            preparedStatement.setString(1, wordToBeUpdated);

            long updateCount = preparedStatement.executeUpdate();

            connection.close();

            return updateCount > 0;
        } catch (SQLException e) {
            throw e;
        }
    }


    public boolean deleteUserWordEntry(String word) throws SQLException {
        try (Connection connection = hikariDataSource.getConnection()) {
            PreparedStatement preparedStatement = connection.prepareStatement("DELETE FROM user_word_entry WHERE word = ?;");
            preparedStatement.setString(1, word);

            long updateCount = preparedStatement.executeUpdate();

            connection.close();

            return updateCount > 0;
        } catch (SQLException e) {
            throw e;
        }
    }

    public UserWordEntry queryWordEntryByWord(String word) throws SQLException {
        try (Connection connection = hikariDataSource.getConnection()) {
            PreparedStatement preparedStatement = connection.prepareStatement("SELECT word, definition, log_absolute_mastery, time_created, time_edited, count_viewed, count_tested, count_tested_correct FROM user_word_entry WHERE word = ?;");
            preparedStatement.setString(1, word);

            ResultSet resultSet = preparedStatement.executeQuery();
            if (!resultSet.next()) return null;
            UserWordEntry userWordEntry = new UserWordEntry(resultSet.getString(1),
                    resultSet.getString(2),
                    resultSet.getDouble(3),
                    resultSet.getLong(4),
                    resultSet.getLong(5),
                    resultSet.getInt(6),
                    resultSet.getInt(7),
                    resultSet.getInt(8));

            connection.close();

            return userWordEntry;
        } catch (SQLException e) {
            throw e;
        }
    }

    public WordExam queryWordExam() throws SQLException {
        try (Connection connection = hikariDataSource.getConnection()) {
            PreparedStatement selectWordPreparedStatement = connection.prepareStatement("SELECT word, definition FROM user_word_entry ORDER BY log_absolute_mastery ASC LIMIT 1;");

            ResultSet resultSet = selectWordPreparedStatement.executeQuery();
            if (!resultSet.next()) return null;
            String word = resultSet.getString(1),
                    definition = resultSet.getString(2);

            PreparedStatement selectChoicesPreparedStatement = connection.prepareStatement("SELECT definition FROM user_word_entry WHERE word != ?  ORDER BY rand() LIMIT 3;");
            selectChoicesPreparedStatement.setString(1, word);
            resultSet = selectChoicesPreparedStatement.executeQuery();
            String definitionChoices[] = new String[4];
            byte correctChoice = Utils.randomIn4();
            if (!resultSet.next()) return null;
            for (int i = 0, j = 0; i < 4; i++)
                if (i == correctChoice)
                    definitionChoices[i] = definition;
                else {
                    definitionChoices[i] = resultSet.getString(1);
                    j++;
                    if (!resultSet.next() && j < 3)
                        return null;
                }

            connection.close();

            return new WordExam(word, definitionChoices, correctChoice);
        } catch (SQLException e) {
            throw e;
        }
    }

    public DefinitionExam queryDefinitionExam() throws SQLException {
        try (Connection connection = hikariDataSource.getConnection()) {
            PreparedStatement selectWordPreparedStatement = connection.prepareStatement("SELECT word, definition FROM user_word_entry ORDER BY log_absolute_mastery ASC LIMIT 1;");

            ResultSet resultSet = selectWordPreparedStatement.executeQuery();
            if (!resultSet.next()) return null;
            String word = resultSet.getString(1),
                    definition = resultSet.getString(2);

            PreparedStatement selectChoicesPreparedStatement = connection.prepareStatement("SELECT word FROM user_word_entry WHERE word != ? ORDER BY rand() LIMIT 3;");
            selectChoicesPreparedStatement.setString(1, word);
            resultSet = selectChoicesPreparedStatement.executeQuery();
            String wordChoices[] = new String[4];
            byte correctChoice = Utils.randomIn4();
            if (!resultSet.next()) return null;
            for (int i = 0, j = 0; i < 4; i++)
                if (i == correctChoice)
                    wordChoices[i] = word;
                else {
                    wordChoices[i] = resultSet.getString(1);
                    j++;
                    if (!resultSet.next() && j < 3)
                        return null;
                }

            connection.close();

            return new DefinitionExam(word, definition, wordChoices, correctChoice);
        } catch (SQLException e) {
            throw e;
        }
    }
}

File: src/jvmMain/kotlin/fancywords/database/FancyWordsDatabase.kt
package fancywords.database

import com.zaxxer.hikari.HikariDataSource
import fancywords.*
import java.sql.SQLException
import java.util.*

class FancyWordsDatabase @Throws(SQLException::class) constructor() : AutoCloseable {
    companion object {
        internal val RECOMMENDED_SIZE = 2 * Runtime.getRuntime().availableProcessors() + 1
        private val DATABASE_NAME = "fancy_words_server"
        private val ALLOW_MULTI_QUERIES_TRUE = "allowMultiQueries=true"
        private val URL = "jdbc:mariadb://localhost:3306/$DATABASE_NAME?$ALLOW_MULTI_QUERIES_TRUE"
        private val USERNAME = "fancywordsserver"
        private val PASSWORD = "xhqycqyys"
    }

    val hikariDataSource = HikariDataSource().apply {
        driverClassName = "org.mariadb.jdbc.Driver"
        //hikariDataSource.setDataSourceClassName("org.mariadb.jdbc.MariaDbDataSource");
        jdbcUrl = URL
        username = USERNAME
        password = PASSWORD

        maximumPoolSize = RECOMMENDED_SIZE

        addDataSourceProperty("prepStmtCacheSize", 256)
        addDataSourceProperty("prepStmtCacheSqlLimit", 2048)
        addDataSourceProperty("cachePrepStmts", true)
        addDataSourceProperty("useServerPrepStmts", true)
    }

    override fun close() =
        hikariDataSource.close()

    fun authenticateUser(username: String, encryptedPassword: ByteArray): Boolean {
        // TODO implement here
        return false
    }

    @Throws(SQLException::class)
    fun queryAllEntries(): List<BriefUserWordEntry> {
        try {
            hikariDataSource.connection.use { connection ->
                val preparedStatement =
                    connection.prepareStatement("SELECT word, definition, log_absolute_mastery FROM user_word_entry ORDER BY word ASC;")

                val resultSet = preparedStatement.executeQuery()
                val entries = ArrayList<BriefUserWordEntry>()
                var i = 0
                while (resultSet.next()) {
                    entries.add(
                        BriefUserWordEntry(
                            resultSet.getString(1),
                            resultSet.getString(2),
                            resultSet.getDouble(3)
                        )
                    )
                    i++
                }

                connection.close()

                return entries
            }
        } catch (e: SQLException) {
            throw e
        }

    }

    @Throws(SQLException::class)
    fun insertUserWordEntry(userWordEntry: UserWordEntry): Boolean {
        try {
            hikariDataSource.connection.use { connection ->
                val preparedStatement =
                    connection.prepareStatement("INSERT INTO user_word_entry VALUES (?, ?, ?, ?, ?, ?, ?, ?);")
                preparedStatement.setString(1, userWordEntry.word)
                preparedStatement.setString(2, userWordEntry.definition)
                preparedStatement.setDouble(3, userWordEntry.logAbsoluteMastery)
                preparedStatement.setLong(4, userWordEntry.timeCreated)
                preparedStatement.setLong(5, userWordEntry.timeEdited)
                preparedStatement.setInt(6, userWordEntry.countViewed)
                preparedStatement.setInt(7, userWordEntry.countTested)
                preparedStatement.setInt(8, userWordEntry.countTestedCorrect)
                val updateCount: Long
                try {
                    updateCount = preparedStatement.executeUpdate().toLong()
                } catch (e: SQLException) {
                    return if (e.errorCode == 1062)
                        false
                    else
                        throw e
                }

                connection.close()

                return updateCount > 0
            }
        } catch (e: SQLException) {
            throw e
        }

    }

    @Throws(SQLException::class)
    fun updateDefinition(wordToBeUpdated: String, definition: String): Boolean {
        try {
            hikariDataSource.connection.use { connection ->
                val preparedStatement =
                    connection.prepareStatement("UPDATE user_word_entry SET definition = ? WHERE word = ?;")
                preparedStatement.setString(1, definition)
                preparedStatement.setString(2, wordToBeUpdated)

                val updateCount = preparedStatement.executeUpdate().toLong()

                connection.close()

                return updateCount > 0
            }
        } catch (e: SQLException) {
            throw e
        }

    }

    @Throws(SQLException::class)
    fun updateMasteryByOne(wordToBeUpdated: String, time: Long): Boolean {
        try {
            hikariDataSource.connection.use { connection ->
                connection.autoCommit = false
                val selectPreparedStatement =
                    connection.prepareStatement("SELECT log_absolute_mastery FROM user_word_entry WHERE word = ?;")
                selectPreparedStatement.setString(1, wordToBeUpdated)
                val resultSet = selectPreparedStatement.executeQuery()
                val logAbsoluteMastery: Double
                if (resultSet.next())
                    logAbsoluteMastery = resultSet.getDouble(1)
                else
                    return false

                val updatePreparedStatement =
                    connection.prepareStatement("UPDATE user_word_entry SET log_absolute_mastery = ? WHERE word = ?;")
                updatePreparedStatement.setDouble(1, increaseLogAbsoluteMasteryByOneByTime(logAbsoluteMastery, time))
                updatePreparedStatement.setString(2, wordToBeUpdated)

                val updateCount = updatePreparedStatement.executeUpdate().toLong()

                connection.commit()
                connection.autoCommit = true

                connection.close()

                return updateCount > 0
            }
        } catch (e: SQLException) {
            throw e
        }

    }

    @Throws(SQLException::class)
    fun updateTimeEdited(wordToBeUpdated: String, timeEdited: Long): Boolean {
        try {
            hikariDataSource.connection.use { connection ->
                val preparedStatement =
                    connection.prepareStatement("UPDATE user_word_entry SET time_edited = ? WHERE word = ?;")
                preparedStatement.setLong(1, timeEdited)
                preparedStatement.setString(2, wordToBeUpdated)

                val updateCount = preparedStatement.executeUpdate().toLong()

                connection.close()

                return updateCount > 0
            }
        } catch (e: SQLException) {
            throw e
        }

    }

    /**
     * This method automatically increase the number of times this word has been viewed by ONE.
     *
     * @param wordToBeUpdated
     * @return
     * @throws SQLException
     */
    @Throws(SQLException::class)
    fun updateCountViewedByOne(wordToBeUpdated: String): Boolean {
        try {
            hikariDataSource.connection.use { connection ->
                val preparedStatement =
                    connection.prepareStatement("UPDATE user_word_entry SET count_viewed = count_viewed + 1 WHERE word = ?;")
                preparedStatement.setString(1, wordToBeUpdated)

                val updateCount = preparedStatement.executeUpdate().toLong()

                connection.close()

                return updateCount > 0
            }
        } catch (e: SQLException) {
            throw e
        }

    }

    @Throws(SQLException::class)
    fun updateCountTested(wordToBeUpdated: String, correct: Boolean): Boolean {
        try {
            hikariDataSource.connection.use { connection ->
                val preparedStatement = connection.prepareStatement(
                    if (correct)
                        "UPDATE user_word_entry SET count_tested = count_tested + 1 AND count_tested_correct = count_tested_correct + 1 WHERE word = ?;"
                    else
                        "UPDATE user_word_entry SET count_tested = count_tested + 1 WHERE word = ?;"
                )
                preparedStatement.setString(1, wordToBeUpdated)

                val updateCount = preparedStatement.executeUpdate().toLong()

                connection.close()

                return updateCount > 0
            }
        } catch (e: SQLException) {
            throw e
        }

    }


    @Throws(SQLException::class)
    fun deleteUserWordEntry(word: String): Boolean {
        try {
            hikariDataSource.connection.use { connection ->
                val preparedStatement = connection.prepareStatement("DELETE FROM user_word_entry WHERE word = ?;")
                preparedStatement.setString(1, word)

                val updateCount = preparedStatement.executeUpdate().toLong()

                connection.close()

                return updateCount > 0
            }
        } catch (e: SQLException) {
            throw e
        }

    }

    @Throws(SQLException::class)
    fun queryWordEntryByWord(word: String): UserWordEntry? {
        try {
            hikariDataSource.connection.use { connection ->
                val preparedStatement =
                    connection.prepareStatement("SELECT word, definition, log_absolute_mastery, time_created, time_edited, count_viewed, count_tested, count_tested_correct FROM user_word_entry WHERE word = ?;")
                preparedStatement.setString(1, word)

                val resultSet = preparedStatement.executeQuery()
                if (!resultSet.next()) return null
                val userWordEntry = UserWordEntry(
                    resultSet.getString(1),
                    resultSet.getString(2),
                    resultSet.getDouble(3),
                    resultSet.getLong(4),
                    resultSet.getLong(5),
                    resultSet.getInt(6),
                    resultSet.getInt(7),
                    resultSet.getInt(8)
                )

                connection.close()

                return userWordEntry
            }
        } catch (e: SQLException) {
            throw e
        }

    }

    @Throws(SQLException::class)
    fun queryWordExam(): WordExam? {
        try {
            hikariDataSource.connection.use { connection ->
                val selectWordPreparedStatement =
                    connection.prepareStatement("SELECT word, definition FROM user_word_entry ORDER BY log_absolute_mastery ASC LIMIT 1;")

                var resultSet = selectWordPreparedStatement.executeQuery()
                if (!resultSet.next()) return null
                val word = resultSet.getString(1)
                val definition = resultSet.getString(2)

                val selectChoicesPreparedStatement =
                    connection.prepareStatement("SELECT definition FROM user_word_entry WHERE word != ?  ORDER BY rand() LIMIT 3;")
                selectChoicesPreparedStatement.setString(1, word)
                resultSet = selectChoicesPreparedStatement.executeQuery()
                val definitionChoices = arrayOfNulls<String>(4)
                val correctChoice = randomIn4()
                if (!resultSet.next()) return null
                var i = 0
                var j = 0
                while (i < 4) {
                    if (i == correctChoice.toInt())
                        definitionChoices[i] = definition
                    else {
                        definitionChoices[i] = resultSet.getString(1)
                        j++
                        if (!resultSet.next() && j < 3)
                            return null
                    }
                    i++
                }

                connection.close()

                return WordExam(word, definitionChoices, correctChoice)
            }
        } catch (e: SQLException) {
            throw e
        }

    }

    @Throws(SQLException::class)
    fun queryDefinitionExam(): DefinitionExam? {
        try {
            hikariDataSource.connection.use { connection ->
                val selectWordPreparedStatement =
                    connection.prepareStatement("SELECT word, definition FROM user_word_entry ORDER BY log_absolute_mastery ASC LIMIT 1;")

                var resultSet = selectWordPreparedStatement.executeQuery()
                if (!resultSet.next()) return null
                val word = resultSet.getString(1)
                val definition = resultSet.getString(2)

                val selectChoicesPreparedStatement =
                    connection.prepareStatement("SELECT word FROM user_word_entry WHERE word != ? ORDER BY rand() LIMIT 3;")
                selectChoicesPreparedStatement.setString(1, word)
                resultSet = selectChoicesPreparedStatement.executeQuery()
                val wordChoices = arrayOfNulls<String>(4)
                val correctChoice = randomIn4()
                if (!resultSet.next()) return null
                var i = 0
                var j = 0
                while (i < 4) {
                    if (i == correctChoice.toInt())
                        wordChoices[i] = word
                    else {
                        wordChoices[i] = resultSet.getString(1)
                        j++
                        if (!resultSet.next() && j < 3)
                            return null
                    }
                    i++
                }

                connection.close()

                return DefinitionExam(word, definition, wordChoices, correctChoice)
            }
        } catch (e: SQLException) {
            throw e
        }

    }
}